### PR TITLE
insights: integrate error retries into new backfiller

### PR DESCRIPTION
**Pull request by coury-clark**
Integrates the retry iterator into the backfiller. This currently just uses the queue retry mechanism to start over, but future changes will build that in explicitly.

## Test plan

Logs from an all repo backfill where I randomly injected errors. After the job was terminally failed I removed the random error injection and retried it once more and it completed successfully, moving into the completed terminal state.

```
[         worker] INFO worker.insights-job.background workerutil/worker.go:332 Dequeued record for processing {"name": "backfill_new_backfill_worker", "id": 36}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_new_handler.go:84 newBackfillHandler called {"handle": {"recordId": 36}}
[         worker] INFO worker.insights-job.background workerutil/worker.go:332 Dequeued record for processing {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "a6a417d552f5cab8", "name": "backfill_in_progress_worker", "id": 37}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:85 inProgressHandler called {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"recordId": 37}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:149 starting primary loop {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"series_id": 19, "backfill_id": 19}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"repo_id": 1}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"repo_id": 2}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"repo_id": 3}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"repo_id": 4}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"repo_id": 5}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"repo_id": 6}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"repo_id": 7}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:154 starting retry loop {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"series_id": 19, "backfill_id": 19}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"repo_id": 5}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"repo_id": 1}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"repo_id": 2}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"repo_id": 3}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "handle": {"repo_id": 4}}
[         worker] WARN worker.insights-job.background.worker.Handle workerutil/worker.go:413 Marked record as errored {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "error": "incomplete backfill"}
[         worker] ERROR worker.insights-job.background.worker.Handle workerutil/worker.go:386 operation.error {"TraceId": "e11ef565eaf2e938edf17c4b9629ded4", "SpanId": "88245eb1da572e2a", "count": 1, "elapsed": 14.1009395, "error": "incomplete backfill"}

[         worker] INFO worker.insights-job.background workerutil/worker.go:332 Dequeued record for processing {"TraceId": "8f4a868ba9ebec21d659510a0e2a0859", "SpanId": "7db20423e1ad2379", "name": "backfill_in_progress_worker", "id": 37}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:85 inProgressHandler called {"TraceId": "8f4a868ba9ebec21d659510a0e2a0859", "SpanId": "892fa22c2a0f5a1f", "handle": {"recordId": 37}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:149 starting primary loop {"TraceId": "8f4a868ba9ebec21d659510a0e2a0859", "SpanId": "892fa22c2a0f5a1f", "handle": {"series_id": 19, "backfill_id": 19}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:154 starting retry loop {"TraceId": "8f4a868ba9ebec21d659510a0e2a0859", "SpanId": "892fa22c2a0f5a1f", "handle": {"series_id": 19, "backfill_id": 19}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "8f4a868ba9ebec21d659510a0e2a0859", "SpanId": "892fa22c2a0f5a1f", "handle": {"repo_id": 2}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "8f4a868ba9ebec21d659510a0e2a0859", "SpanId": "892fa22c2a0f5a1f", "handle": {"repo_id": 3}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "8f4a868ba9ebec21d659510a0e2a0859", "SpanId": "892fa22c2a0f5a1f", "handle": {"repo_id": 4}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "8f4a868ba9ebec21d659510a0e2a0859", "SpanId": "892fa22c2a0f5a1f", "handle": {"repo_id": 5}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "8f4a868ba9ebec21d659510a0e2a0859", "SpanId": "892fa22c2a0f5a1f", "handle": {"repo_id": 1}}
[         worker] WARN worker.insights-job.background.worker.Handle workerutil/worker.go:413 Marked record as errored {"TraceId": "8f4a868ba9ebec21d659510a0e2a0859", "SpanId": "892fa22c2a0f5a1f", "error": "incomplete backfill"}
[         worker] ERROR worker.insights-job.background.worker.Handle workerutil/worker.go:386 operation.error {"TraceId": "8f4a868ba9ebec21d659510a0e2a0859", "SpanId": "892fa22c2a0f5a1f", "count": 1, "elapsed": 5.860580583, "error": "incomplete backfill"}

[         worker] INFO worker.insights-job.background workerutil/worker.go:332 Dequeued record for processing {"TraceId": "ad22242a87c13ead55cf808993304ad4", "SpanId": "963727364d082744", "name": "backfill_in_progress_worker", "id": 37}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:85 inProgressHandler called {"TraceId": "ad22242a87c13ead55cf808993304ad4", "SpanId": "d81c0e6113b238ca", "handle": {"recordId": 37}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:149 starting primary loop {"TraceId": "ad22242a87c13ead55cf808993304ad4", "SpanId": "d81c0e6113b238ca", "handle": {"series_id": 19, "backfill_id": 19}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:154 starting retry loop {"TraceId": "ad22242a87c13ead55cf808993304ad4", "SpanId": "d81c0e6113b238ca", "handle": {"series_id": 19, "backfill_id": 19}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "ad22242a87c13ead55cf808993304ad4", "SpanId": "d81c0e6113b238ca", "handle": {"repo_id": 4}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "ad22242a87c13ead55cf808993304ad4", "SpanId": "d81c0e6113b238ca", "handle": {"repo_id": 5}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "ad22242a87c13ead55cf808993304ad4", "SpanId": "d81c0e6113b238ca", "handle": {"repo_id": 2}}
[         worker] WARN worker.insights-job.background.worker.Handle workerutil/worker.go:413 Marked record as errored {"TraceId": "ad22242a87c13ead55cf808993304ad4", "SpanId": "d81c0e6113b238ca", "error": "incomplete backfill"}
[         worker] ERROR worker.insights-job.background.worker.Handle workerutil/worker.go:386 operation.error {"TraceId": "ad22242a87c13ead55cf808993304ad4", "SpanId": "d81c0e6113b238ca", "count": 1, "elapsed": 3.467747459, "error": "incomplete backfill"}


-------------- After manual restart (since job was terminally failed at this point) and removing forced errors ---------


[         worker] INFO worker.insights-job.background workerutil/worker.go:332 Dequeued record for processing {"TraceId": "d5bf39386e80762aba1bb4fec3033d7b", "SpanId": "65299a90a6f42e88", "name": "backfill_in_progress_worker", "id": 37}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:85 inProgressHandler called {"TraceId": "d5bf39386e80762aba1bb4fec3033d7b", "SpanId": "adde91c508feb0c6", "handle": {"recordId": 37}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:149 starting primary loop {"TraceId": "d5bf39386e80762aba1bb4fec3033d7b", "SpanId": "adde91c508feb0c6", "handle": {"series_id": 19, "backfill_id": 19}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:154 starting retry loop {"TraceId": "d5bf39386e80762aba1bb4fec3033d7b", "SpanId": "adde91c508feb0c6", "handle": {"series_id": 19, "backfill_id": 19}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "d5bf39386e80762aba1bb4fec3033d7b", "SpanId": "adde91c508feb0c6", "handle": {"repo_id": 2}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "d5bf39386e80762aba1bb4fec3033d7b", "SpanId": "adde91c508feb0c6", "handle": {"repo_id": 4}}
[         worker] INFO worker.insights-job.background.worker.Handle scheduler/backfill_state_inprogress_handler.go:133 doing iteration work {"TraceId": "d5bf39386e80762aba1bb4fec3033d7b", "SpanId": "adde91c508feb0c6", "handle": {"repo_id": 5}}

```

### CODE REVIEW

File: enterprise/internal/insights/scheduler/iterator/repo_iterator.go [hunk 1]
```diff
@@ -192,6 +192,22 @@ func (p *PersistentRepoIterator) MarkComplete(ctx context.Context, store *basest
 	return nil
 }
 
+func (p *PersistentRepoIterator) IsComplete() bool {
+	return p.PercentComplete == 1
```
Comment (chwarwick) [inline enterprise/internal/insights/scheduler/iterator/repo_iterator.go:196]: this is a float I can never remember if this is going to be an issue or not.

File: enterprise/internal/insights/scheduler/backfill_state_inprogress_handler.go [hunk 2]
```diff
@@ -101,52 +105,83 @@ func (h *inProgressHandler) Handle(ctx context.Context, logger log.Logger, recor
...
 		},
 	}); err != nil {
-		return err
+		return errors.Wrap(err, "InProgressHandler.SetInsightSeriesRecordingTimes")
 	}
 
-	// todo handle errors down here after the main loop https://github.com/sourcegraph/sourcegraph/issues/42724
+	type nextFunc func() (api.RepoID, bool, iterator.FinishFunc)
+	itrLoop := func(nextFunc nextFunc) error {
+		for true {
```
Comment (chwarwick) [inline enterprise/internal/insights/scheduler/backfill_state_inprogress_handler.go]:
```suggestion
		for {
```

**Review by chwarwick — APPROVED** [review]
Looks good just some questions about errors, looks like the `TotalErrors` on the repo iterator is a bit of a strange value.  It's the sum of the attempts for repos that errored?  So if after the 3? attempts there are 3 remaining errors it's going to report 9?  ErroredRepos feels like the better value we may want to remove that `TotalErrors` it's kind of misleading in that it's FailedRepos * retries.

File: enterprise/internal/insights/scheduler/iterator/repo_iterator.go (hunk 1 above)
Comment (coury-clark) [reply enterprise/internal/insights/scheduler/iterator/repo_iterator.go]: Yeah, I was skeptical of that so I'm going to just change it [here](https://github.com/sourcegraph/sourcegraph/pull/43816/commits/35886151abd447c1b21181b0da45e782f7c92c58)

File: enterprise/internal/insights/scheduler/backfill_state_inprogress_handler.go (hunk 2 above)
Comment (coury-clark) [reply enterprise/internal/insights/scheduler/backfill_state_inprogress_handler.go]: Linter yelled at me too, should be fixed

**Comment by coury-clark** [comment]
@chwarwick Just for the purposes of having some useful information I'm writing it in a log. I'm not married to the idea of either of these, which is why the log includes both.

**Pull request merged**